### PR TITLE
Entity storage

### PR DESCRIPTION
**Pull request by AlexanderLitus**

### CODE REVIEW

**Comment by AlexanderLitus** [comment]
Almost done.

File: .idea/copyright/TeamDev_Open_Source.xml [hunk 1]
```diff
@@ -0,0 +1,7 @@
...
+    <option name="allowReplaceKeyword" value="2000-" />
+    <option name="myName" value="TeamDev Open-Source" />
+    <option name="notice" value="Copyright 2015, TeamDev Ltd. All rights reserved.&#10;&#10;Redistribution and use in source and/or binary forms, with or without&#10;modification, must retain the above copyright notice and the following&#10;disclaimer.&#10;&#10;THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS&#10;&quot;AS IS&quot; AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT&#10;LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR&#10;A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT&#10;OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,&#10;SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT&#10;LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,&#10;DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY&#10;THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT&#10;(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE&#10;OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE." />
```
Comment (alexander-yevsyukov) [inline .idea/copyright/TeamDev_Open_Source.xml]: 2016

File: hsqldb/src/main/java/org/spine3/server/storage/hsqldb/HsqlStorageFactory.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+     * NOTE: {@code dataSourceClassName} config property is set automatically.
+     *
+     * <p>Examples of JDBC URL:
+     *
+     * <p>{@code jdbc:hsqldb:hsql://localhost:9001/dbname;ifexists=true}
+     * <p>{@code jdbc:hsqldb:mem:inmemorydb} (for in-memory database)
+     *
+     * @param config the config used to create {@link HikariDataSource}.
+     */
+    public static HsqlStorageFactory newInstance(HikariConfig config) {
```
Comment (alexander-yevsyukov) [inline hsqldb/src/main/java/org/spine3/server/storage/hsqldb/HsqlStorageFactory.java]: This way we exposes Hikari as part of our public API. This must be avoided.

File: hsqldb/src/main/java/org/spine3/server/storage/hsqldb/HsqlEntityStorage.java [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+import static org.spine3.protobuf.Messages.toAny;
+import static org.spine3.util.Identifiers.idToString;
+
+/**
+ * The implementation of the entity storage based on HyperSQL Database.
+ *
+ * @see HsqlStorageFactory
+ * @author Alexander Litus
+ */
+@SuppressWarnings("DuplicateStringLiteralInspection")
```
Comment (alexander-yevsyukov) [inline hsqldb/src/main/java/org/spine3/server/storage/hsqldb/HsqlEntityStorage.java]: Don't suppress warnings like this class-wide. What strings are duplicated. Perhaps, it's the sign of a design need?

File: hsqldb/src/main/java/org/spine3/server/storage/hsqldb/HsqlEntityStorage.java [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+class HsqlEntityStorage<I, M extends Message> extends EntityStorage<I, M> implements AutoCloseable {
```
Comment (alexander-yevsyukov) [inline hsqldb/src/main/java/org/spine3/server/storage/hsqldb/HsqlEntityStorage.java]: What details of this entity storage are specific to HSQL?
How would look an entity storager for MySQL, or for Oracle?

File: hsqldb/src/test/java/org/spine3/server/storage/hsqldb/HsqlEntityStorageShould.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+    @AfterClass
+    public static void tearDownClass() {
+        DATABASE.execute(SHUTDOWN_SQL);
+        DATABASE.close();
+    }
+
+    @Test
+    public void close() {
```
Comment (alexander-yevsyukov) [inline hsqldb/src/test/java/org/spine3/server/storage/hsqldb/HsqlEntityStorageShould.java]: Why not name it `close_itself()` or `auto_close_`?
The way it's now is a bit confusing. If you overlook `@Test` annotation, which I initially did, the method looks like a real service method of the test case.

File: hsqldb/src/test/java/org/spine3/server/storage/hsqldb/HsqlEntityStorageShould.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    private static final String DROP_TABLE_SQL = "DROP TABLE " + ENTITIES;
+
+    private static final String SHUTDOWN_SQL = "SHUTDOWN";
+
+    public HsqlEntityStorageShould() {
+        super(STORAGE);
+    }
+
+    private static HsqlDb newHsqlDb() {
+        final HikariConfig config = new HikariConfig();
```
Comment (alexander-yevsyukov) [inline hsqldb/src/test/java/org/spine3/server/storage/hsqldb/HsqlEntityStorageShould.java]: Please do hide Hikari from our API.

**Comment by AlexanderLitus** [comment]
@alexander-yevsyukov, PTAL.

File: rdbms/src/main/java/org/spine3/server/storage/rdbms/DataSourceConfig.java [hunk 7]
```diff
@@ -0,0 +1,505 @@
...
+        private Long connectionTimeout;
+        private Long idleTimeout;
+        private Long maxLifetime;
+        private String connectionTestQuery;
+        private Integer maxPoolSize;
+        private String poolName;
+
+        @SuppressWarnings("MethodWithMoreThanThreeNegations")
+        public DataSourceConfig build() {
+            // Either dataSourceClassName or jdbcUrl is required
```
Comment (alexander-yevsyukov) [inline rdbms/src/main/java/org/spine3/server/storage/rdbms/DataSourceConfig.java]: This should be in the message, not just in comments.

File: rdbms/src/main/java/org/spine3/server/storage/rdbms/DataSourceConfig.java [hunk 8]
```diff
@@ -0,0 +1,505 @@
...
+         * Required
+         */
+        private String dataSourceClassName;
+        private String jdbcUrl;
+        private String username;
+        private String password;
+
+        /**
+         * Optional
+         */
```
Comment (alexander-yevsyukov) [inline rdbms/src/main/java/org/spine3/server/storage/rdbms/DataSourceConfig.java]: Have nullable fields annotated as such.

File: rdbms/src/main/java/org/spine3/server/storage/rdbms/package-info.java [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+@ParametersAreNonnullByDefault
+package org.spine3.server.storage.rdbms;
```
Comment (alexander-yevsyukov) [inline rdbms/src/main/java/org/spine3/server/storage/rdbms/package-info.java]: JDBC assumes we connect to a SQL database. This in turn means that the underlying database is relational. Therefore, `jdbc` as a package name would be easier to understand.

File: rdbms/src/main/java/org/spine3/server/storage/rdbms/JdbcEntityStorage.java [hunk 10]
```diff
@@ -0,0 +1,295 @@
...
+        try (ConnectionWrapper connection = dataSource.getConnection(true);
+             PreparedStatement statement = connection.prepareStatement(sql)) {
+            statement.execute();
+        } catch (SQLException e) {
+            log().error("Error during table creation, table name = " + tableName, e);
+            throw new DatabaseException(e);
+        }
+    }
+
+    private boolean containsEntity(String id, ConnectionWrapper connection) {
```
Comment (alexander-yevsyukov) [inline rdbms/src/main/java/org/spine3/server/storage/rdbms/JdbcEntityStorage.java]: Have `connection` as the first parameter for all methods. We need to be consistent across the whole API.

**Comment by AlexanderLitus** [comment]
@alexander-yevsyukov, PTAL.

**Pull request merged**